### ISSUE TITLE
Verilog-A modules builder concept

### ISSUE THREAD

**Issue by ra3xdh** [Qucs/qucs#318]
I would like to present my concept of Verilog-A modules builder. Test branch https://github.com/ra3xdh/qucs/tree/verilog-A contains source code. Verilog-A modules builder allows you to convert a Qucs subciruit (_.sch) to Verilog-A module (_.va). This concept was discussed with @MikeBrinson  about a month ago. At current state only EDDs and resistors are supported. There are no error-checks in schematic. Of course, there may be bugs.

How to use:
- Checkout and build branch
- Run Qucs, open subcircuit (*.sch)
- EDIT: Click main menu _File->Build Verilog-A module from subcircuit_
- Select filename and location to save module
- View module with test editor

Here is test circuit: https://gist.github.com/ra3xdh/0a579afdb7c24c003a79 . 

It is tunnel diode model.
![tunn](https://cloud.githubusercontent.com/assets/4920080/8766311/f501a586-2e3b-11e5-9c80-0a33ad228283.png)

And here is auto-generated Verilog-A code for it. 

```
`include "disciplines.vams"
`include "constants.vams"
module tunnsub(K, A);
inout K, A;
electrical _net0, K, A;
parameter real Is=1e-12;
parameter real Ip=1e-5;
parameter real Iv=1e-6;
parameter real Vp=0.1;
parameter real Vv=0.4;
parameter real K=5;
parameter real C=0.01p;
parameter real Temp0=300;
real VT;
analog begin 
@(initial model)
begin 
VT=(`P_K*Temp0)/`P_Q;
end
I(_net0,K) <+ Is*(exp(V(_net0,K)/VT)-1.0);
I(_net0,K) <+ ddt(C*V(_net0,K));
I(_net0,K) <+ Iv*exp(K*(V(_net0,K)-Vv));
I(_net0,K) <+ Ip*(V(_net0,K)/Vp)*exp((Vp-V(_net0,K))/Vp);
V(_net0,A) <+ I(_net0,A)*(1e-3);
end
endmodule
```

**Comment by felix-salfelder** [comment]
this looks fun! it is consistent with #316. the code should go next to `verilog.cpp`, where the verilog netlister is implemented (or even share some code?).

> Click main menu Simulation->Build Verilog-A module from subcircuit

no. what we need is File->Export as.. and in the filetype dropdown menu, we should list the language extensions. #316 only provides a CLI option right now. the GUI thing should be implemented independently.

> V(_net0,A) <+ I(_net0,A)*(1e-3);

verilog-A supports subdevices. i'm inclined to propose that an export to verilog-A should faithfully reproduces subdevices, and not all subdevices can be easily expressed as a branch contribution. a user should choose an equation component instead of a subdevice if (s)he needs a branch contribution statement in the exported module.

**Comment by guitorri** [comment]
@ra3xdh does this code has to be after spice4qucs?
I believe the `V() <+` contributions are not currently supported by the ADMS/qucsator interface...

**Comment by ra3xdh** [comment]
@guitorri Yes, this branch is fork of `rebase_spice4qucs`. Development of Verilog-A modules builder and SPICE-support stuff will go in parallels.

> I believe the V() <+ contributions are not currently supported by the ADMS/qucsator interface...

No problems, I can use current-oriented form `I() <+` of equations instead. I saw such equations in `src/components/verilog` models used by Qucs via ADMS.

@felix-salfelder 

> no. what we need is File->Export as.. and in the filetype dropdown menu
Yes, I will move verilog-A exporter to File menu  group.

> verilog-A supports subdevices

I know about subdevices. But RCL can use not only constant values. There may be SPICE behavioral RCLs too. I think it's better to export them as equations.

I have plans to support at least the following components:
* EDDs
* RCL (usual and behavioral )
* SPICE B-type sources
* DC sources

These components are most used in compact models.

**Comment by felix-salfelder** [comment]
> I know about subdevices. But RCL can use not only constant values. There may be SPICE behavioral RCLs too. I think it's better to export them as equations.

an `R` (same with `C` and `L` and some others) with a non-constant value is no longer a component of type `R`, which corresponds to a (constant) resistor in QUCS (and in verilog). it does not exist in QUCS, and it *must* not exist. unlike SPICE, QUCS uses different names for different devices. a QUCS device is much like a verilog device, fully specified by its type name. it is not a big deal to add a different component for different functionality. think of `VRect`, `VPulse` -- these work perfectly well, are not hacks, easy to understand, and still can be easily exported to some SPICE dialect *AND* to all other languages.

tl;dr:
* QUCS does not have to become SPICE to provide components that SPICE uses,
* QUCS must not become SPICE to maintain sanity.

we need standard compliance more than anything else, there's no point in moving further away for the sake of a particular and obsolete corner case (such as SPICE). don't pull in weird hacks here. these will break the future transition to verilog (or maybe the transition will break your code and the schematics for all the users).

thanks

**Comment by ra3xdh** [comment]
> * QUCS does not have to become SPICE to provide components that SPICE uses,
* QUCS must not become SPICE to maintain sanity.

SPICE remains an industrial standard. And we should follow it despite our wishes. Many skilled engineers asked me about full SPICE compatibility in Qucs. Discussion at #275 will be continued later.

**Comment by felix-salfelder** [comment]
i wasn't aware that i argued against spice *compatibility*. let me clarify.

the standard that is worth following is verilog, this is the one, the industry is using or heading towards to. QUCS should (for the sake of it) support both, verilog and spice, this is postulated by the U in the name. by poisoning the core with SPICE semantics (such as device "variants"), you thwart future development, and sane interoperability -- for nothing.

i have explained previously, how supporting SPICE does not interfere with the current component semantics that has been chosen (carefully!) by the original QUCS author. and i truly hope that you won't touch it.

thank you

PS: and please don't use the words "SPICE" and "standard" in a sentence without negation.

**Comment by felix-salfelder** [comment]
as a side remark: what is wrong about letting the user decide how to create his model? that would mean (for the verilog-a netlister)
* user places component => user gets subcircuit component
* user places voltage+EQN => user gets `V(..)<+` statement
* user places current+EQN => user gets `I(..)<+` statement

in addition to that you could still add special SPICE components that behave completely different to that. in *addition*. and with a huge portability warning attached.

**Comment by MikeBrinson** [comment]
The synthesis of Verilog-A code from Qucs EDD behavioural models is important because it fills in one of the missing links in the compact model development cycle.  The Qucs EDD structure has been designed to provide a one-to-one link between EDD and Verilog-A statements, see the Qucs presentations given at the MOS-AK workshops (http://www.mos-ak.org/).  Currently, new models can
be developed using EDD and iteratively tested via Qucs GUI.  However, an equivalent model described in Verilog-A has to be generated by hand !  The new synthesis engine will allow automatic generation of Verilog-A code from EDD, and other Qucs components and subcircuits.  As the ADMS Verilog-A to C++ compiler becomes further developed the generated C++ models will link directly to the Qucs, ngspice, Xyce and gnucap simulation engines.   For this route to fully work, much more work is needed on improving the subset of the Verilog-A HDL allowed by ADMS and also on improving the linking of the generated C++ models to all the above listed circuit simulators.

To assist with the generation of the Verilog-A code from Qucs EDD and other component models I
will post a series of modelling solutions to problems like how does one represent V(n) <+ I(n) and 
nonlinear inductors in Verilog-A.

**Comment by MikeBrinson** [comment]
Hello Vadim
Thanks for taking time to push forward development of the Qucs EDD to Verilog-A synthesiser.After reading all the email on this subject I added a note on the GitHub column pointing out theimportance of this subject for compact modelling development.
Over the coming weeks I will contribute to the development of this topic, adding my knowledge ofthe Qucs EDD element, Verilog-A HDL statements and the current capabilities and limitations of the ADMScompiler software.  The recent release of Xyce 3.0 makes the development of the Qucs EDD to Verilog-A synthesisermost timely.  To make the synthesiser really effective we will probably need to make changes to the ADMS 2.34compiler and the Qucs XML scripts.  One obvious example is how we deal with linear passive components whengenerating Verilog-A code.  Other obvious Verilog-A statements to be dealt with include V(n) <+ 0.0 and V(n) <+ I(n)*R.
To start with I will try using your Verilog-A synthesiser code to generate Verlog-A models for a number of simple and more complexcompact device models. In addition  I will try develop a number of test cases which show the EDD schematics and the resulting generatedVerilog-A code for a number important modelling cases which are not currently implemented by ADMS.

Would you like me to send you directly my notes and test data or to simply enter them on the Verilog-A synthesiser GitHub web site?
Best wishes.
Mike

 Mike Brinson

mbrin72043@yahoo.co.uk

**Comment by felix-salfelder** [comment]
> Other obvious Verilog-A statements to be dealt with include V(n) <+ 0.0 and V(n) <+ I(n)*R.

@MikeBrinson i am working on the left hand side potential contribution in gnucap-adms right now. maybe we should fix a common denominator in the implicit rules. how about the Qucs/ADMS issue tracker?

**Comment by ra3xdh** [comment]
@MikeBrinson , the latest code at https://github.com/ra3xdh/qucs/tree/verilog-A branch contains current-oriented resistor model ( `I(...) <+ V(...)/R` ), capacitor model, and Menu point moved to *File* menu.

**Comment by MikeBrinson** [comment]
The following diagram shows a set of experimental EDD/R/C  "templates" for constructing  more complex experimental compact Verilog-A device models.  These templates can be stored on a Qucs schematic (in a suitable project directory) and copied onto a working schematic when required.

![verilogatemplatemodels](https://cloud.githubusercontent.com/assets/3287115/9137336/c1b58980-3d14-11e5-8baa-85443efa1ddc.png)

These are just a small selection of possible template types which can be easily constructed.

As an example of their use consider the template for an inductance and the example of its use shown in the next diagram:

![inductanceexample](https://cloud.githubusercontent.com/assets/3287115/9137449/85c558dc-3d15-11e5-9ce2-b5852ac63fd3.png)

By combing the L EDD template with one external linear resistor an inductance model can be simply
formed: see next diagram
![lmodel](https://cloud.githubusercontent.com/assets/3287115/9137642/e7972eb8-3d16-11e5-869c-3340e0114da1.png)

Overall testing of the new Verilog-A/R synthesised model is also straight forward - see next diagram.

![testlb1](https://cloud.githubusercontent.com/assets/3287115/9137665/24392b78-3d17-11e5-8328-091d31d4bfa6.png)

**Comment by MikeBrinson** [comment]
The unofficial Qucs releases Qucs-0.0.19S-rc1 and Qucs-0.0.19S-rc2 appear to have a bug in the naming of the circuit  ground node.  Some ground nodes were numbered 0 rather than gnd which in turn causes the ADMS compiler to fail.  I have pushed a block of C++ code to Qucs branch Verilog-A which should correct this problem.  The following test circuit and generated Verilog-A code suggest that this bug has been eliminated:
![eddrctestcircuit](https://cloud.githubusercontent.com/assets/3287115/9138180/2546fac4-3d1a-11e5-8a90-7003d79b0538.png)

Please note the test circuit only generates a list of Verilog-A statements with changing node naming.
It has no other functionality.

![verilogtestcode](https://cloud.githubusercontent.com/assets/3287115/9138188/3159c062-3d1a-11e5-849b-533dd2f659ff.png)

**Comment by MikeBrinson** [comment]
Added thermal current noise to resistors in Verilog-A synthesised modules.

The following example outlines how thermal current noise is handled by Qucs/spice4qucs when
synthesising Verilog-A modules from Qucs subcircuits. 
![thermalnoise](https://cloud.githubusercontent.com/assets/3287115/9155036/2011822a-3ea2-11e5-92ce-5c3fe7c0c7ad.png)

In this example a simple two resistor voltage divider is used to demonstrate the generation, and plotting, of thermal noise current and voltages.

Note how ground node terminals within a subcircuit are handled.  Verilog-A includes statement

ground gnd;

However, with the current Adms/qucs0.0.19 combination  Verilog-A statement "ground" is not fully
implemented.  As a work around  label the ground net, in a subcircuit,  gnd and connect it to a subcircuit interface pin.  This can then be connected to the main earth on a circuit schematic, ensuring that a ground net within a subcircuit is not LEFT FLOATING!  Also not with a high percentage of Verilog-A subcircuits a ground net is not defined.

**Comment by ra3xdh** [comment]
@MikeBrinson  I have added improved fix for `gnd` bug by last commmit at `rebase_spice4qucs` branch. Now SPICE `0` node name should not appear in Verilog-A modules. Added Verilog-A/SPICE independent conversion for node voltages. Please test on your side.

Thank you for thermal noise and inductance models! I will integrate it during the next week.

**Comment by MikeBrinson** [comment]
As an extension to the thermal noise model for resistors I have added models for semiconductor device shot and flicker noise.  The attached diagrams give details of the models and a simple test circuit.  Both the shot and flicker models require, as input, device DC bias current.  In response they output device shot and noise current in the AC domain.  In the test example a DC input current is
swept between 0 and 10A and the noise current measured and plotted as a function of frequency. 
Notice how the VA model takes Verilog-A statements as input to the noise model and outputs a working Verilog-A module.
![test_edd_va_noise_models](https://cloud.githubusercontent.com/assets/3287115/9195527/67d8cbde-401b-11e5-8372-7f6f1c637bea.png)
![test_edd_va_noise](https://cloud.githubusercontent.com/assets/3287115/9195531/6c2d04ac-401b-11e5-9164-e374be5010f5.png)

**Comment by MikeBrinson** [comment]
Added an EDD/Verilog-A CCVS to Qucs/spice4qucs.  This fundamental device is needed in compact device modelling to probe for aDC current in a branch and output a voltage with same magnitude as the probed current.  To test the operation of this device and check the operation of the Qucs/spice4qucs Verilog-A synthesiser a compact model of a basic semiconductor diode was constructed. The next three diagrams give details of the diode model and its operation based on the EDD schematic and the synthesised Verilog-A code model.  Note both forms of diode model give very similar results implying 
![diodefig1](https://cloud.githubusercontent.com/assets/3287115/9254887/e5ec92d0-41dd-11e5-85c8-db8c0b45ea18.png)
![diode_edd](https://cloud.githubusercontent.com/assets/3287115/9254892/eb64f914-41dd-11e5-941d-15975c80159c.png)
![diode_va](https://cloud.githubusercontent.com/assets/3287115/9254895/f31a1450-41dd-11e5-8f5a-344feb0cc548.png)


that the basic Verilog-A syntheiser and the new CCVS model are functioning correctly.

**Comment by MikeBrinson** [comment]
Added CCCS, VCCS and VCVS linear controlled sources.  Test results for EDD and Verilog-A versions show below.
![tlinearcontrolledsources](https://cloud.githubusercontent.com/assets/3287115/9273751/6d0d5796-4287-11e5-8a86-81917ad86104.png)
![testeddsources](https://cloud.githubusercontent.com/assets/3287115/9273754/73e541c8-4287-11e5-83a0-4137402612ae.png)
![testvasources](https://cloud.githubusercontent.com/assets/3287115/9273757/7af3192c-4287-11e5-9799-8c07b7524303.png)

**Comment by MikeBrinson** [comment]
More Verilog-A synthesised diode test results - temperature variation of diode properties and resulting I/V curves for EDD, Verilog-A and Qucs diode models:

![tempdiodefig1](https://cloud.githubusercontent.com/assets/3287115/9279617/9cbf0b78-42b0-11e5-86ac-2d2441b23b2f.png)
![testedddiode3temp](https://cloud.githubusercontent.com/assets/3287115/9279621/a16dd366-42b0-11e5-8798-d89571e00183.png)
![testvadiode3temp](https://cloud.githubusercontent.com/assets/3287115/9279632/a7526b66-42b0-11e5-84d7-986683b6e10b.png)

**Comment by ra3xdh** [comment]
Added inductance model and noise source model suggested by @MikeBrinson . Here is an example of usage:
![ind](https://cloud.githubusercontent.com/assets/4920080/9289310/21ea9eca-4374-11e5-8eda-afd8d4b5ced8.png)
Here is code example:
~~~
`include "disciplines.vams"
`include "constants.vams"
module Lsub(p1, p2);
inout p1, p2;
electrical p2, p1;
parameter real L1=1e-6;
parameter real Rleak=1e6;
analog begin 
I(gnd,_net0L1) <+ ddt(-V(_net0L1))+V(p2,p1); // Inductor: L1 
I(p2,p1) <+ V(_net0L1)/(L1+1e-15);
I(p1,p2) <+ flicker_noise(1e-6,0); // I1 noise source 
I(p1,p2) <+ V(p1,p2)/Rleak; // B1 source
end
endmodule
~~~

**Comment by MikeBrinson** [comment]
More test results.  A long channel EKVv2p6 EDD based model.  This model demonstrates the use of EDD elements for simulating the non-linear DC I/V performance of an n type MOSFET.  The EDD model, synthesised Verilog-A code and typical test results are given in the following diagrams.  Please note that, at this stage in the development cycle,
![ekv_2p6_fig1](https://cloud.githubusercontent.com/assets/3287115/9290212/fbb8cbb0-4381-11e5-9e8b-538e1f5a8510.png)
![ekv_2p6_fig2](https://cloud.githubusercontent.com/assets/3287115/9290213/0132d18a-4382-11e5-9ad6-99f0b27dc26d.png)
![ekv_2p6_edd](https://cloud.githubusercontent.com/assets/3287115/9290215/081581f0-4382-11e5-8118-5e62fd274ac2.png)
![ekv_2p6_va](https://cloud.githubusercontent.com/assets/3287115/9290216/0df8dc16-4382-11e5-98bf-775acec8416c.png)



 the generated Verilog-A code is not optimized for minimum run time.

**Comment by MikeBrinson** [comment]
Attached is an optimised version of the EKV2p6 model:  NOTE (1) I(a,b) <+ 0; statements are now removed by the Verilog-A synthesiser, (2) the 1Ohm and 1e9 Ohm Verilog-A noise contribution statements have been edited out as these shouud NOT contribute to the device noise footprint, and (3) the EDD model has been minimised in size reducing significantly
![ekv_edd_opt_model](https://cloud.githubusercontent.com/assets/3287115/9311568/ed3a3ef0-450e-11e5-9151-ee72cb279964.png)
![ekv_edd_opt_model_verilog-a](https://cloud.githubusercontent.com/assets/3287115/9311574/f5bb0e88-450e-11e5-8257-de76b503c6f6.png)
![ekv_edd_opt](https://cloud.githubusercontent.com/assets/3287115/9311578/fb138784-450e-11e5-96ee-411010005e68.png)
![ekv_va_opt](https://cloud.githubusercontent.com/assets/3287115/9311582/010d11b4-450f-11e5-8e5a-39da05335f89.png)



 the number of contribution statements.

**Comment by MikeBrinson** [comment]
Made improvements to ampere_noise.cpp and ampere_noise.h to allow shot and flicker noise sources to be synthesised from EDD compact device models.  The following diagrams show how to generate shot and flicker noise, their EDD templates and a set of synthesised Verilog-A code for the noise models.

![sfnoisefig1](https://cloud.githubusercontent.com/assets/3287115/9361224/4a7f6052-4692-11e5-8070-f8c36256b7b4.png)
![sandfedd](https://cloud.githubusercontent.com/assets/3287115/9361233/4fb2686c-4692-11e5-8235-092708d98755.png)
![sandfva](https://cloud.githubusercontent.com/assets/3287115/9361238/5a846006-4692-11e5-83e2-11c730639364.png)

**Comment by MikeBrinson** [comment]
The Qucs inductance mode has been extended to include generation of Verilog-A code; see attached
results below for phase amplitude and phase results from a basic series LCR network.  An example of the Verilog-A code generated by the Qucs synthesizer for a multi-terminal LCR network is shown last.
![serieslcr](https://cloud.githubusercontent.com/assets/3287115/9424539/646c34b6-48e7-11e5-93bd-c8a93465a377.png)
![lc_va](https://cloud.githubusercontent.com/assets/3287115/9424540/6d661672-48e7-11e5-9615-e50407a0249b.png)
![lmatrixfig1](https://cloud.githubusercontent.com/assets/3287115/9424543/74f3a42c-48e7-11e5-9c81-e322ccd737ef.png)

**Comment by MikeBrinson** [comment]
Made changes to the resistor, capacitor and inductance models to eliminate internal node references to ground node gnd.  he following test example shows a series of RCL components with different connections to earth.  NOTE this circuit is for test purposes and does not have a specific function.

![testrclgnd](https://cloud.githubusercontent.com/assets/3287115/9450964/c59c3a7a-4aa3-11e5-9fb7-1c91c7fc3d38.png)
![rclgndva](https://cloud.githubusercontent.com/assets/3287115/9450972/caeb0a74-4aa3-11e5-82a1-35a6651f681d.png)

**Comment by MikeBrinson** [comment]
Changed ampere_noise model to remove node gnd references.  The following test circuit and generated Verilog-A listing demonstrates the removal of node gnd.
![vanoisefig2](https://cloud.githubusercontent.com/assets/3287115/9468536/146d5bde-4b3a-11e5-9db8-fb7a21c9cff0.png)
![vanoisefig1](https://cloud.githubusercontent.com/assets/3287115/9468548/1cd76288-4b3a-11e5-9212-2dd7bde353e6.png)

**Comment by MikeBrinson** [comment]
Changed linear controlled source code to remove node ground (gnd) from synthesized Verilog-A modules.  The next two diagrams illustrate the process of ground node removal during Verilog-A code generation by the Qucs synthesizer.
![csfig1](https://cloud.githubusercontent.com/assets/3287115/9492407/1e7479f4-4bef-11e5-8106-dd2fac49c998.png)
![csfig2](https://cloud.githubusercontent.com/assets/3287115/9492409/235c435c-4bef-11e5-8b40-07983989d604.png)

**Comment by ra3xdh** [comment]
Colsing. This Verilog-A module builder still be presented in Qucs-S http://ra3xdh.github.io/ . You can use it until the code will be backported in Qucs using modularity concepts.
